p2p, rpc: Manual block-relay-only connections with addnode

mzumsande opened the pull request:
This implements the suggestion from #23763 to introduce an option to establish block-relay-connections manually with the `-addnode` RPC.
Adding these can make sense for a node operator that wants to be connected to a anonymity networks like Tor or I2P, but also wants to have additional protection against eclipse attacks: Following the best chain can be more of an issue on anonymity networks because these are smaller and it can be easier to create a lot of sybil nodes there.

In that situation, manual block-relay-only connections to peers on clearnet networks can help us staying connected to the best chain, but in contrast to normal manual connections, transactions and addresses aren't transmitted on these links - in particular not our own address or transaction from our wallet. This increases privacy and will also make it harder to perform fingerprinting attacks (connecting our identities over different networks).

Manual Block-Relay connections:
- can be specified with `-addnode` RPC, both with the `add` and `onetry` command
- can be specified with the `-addnode` bitcoind arg (or in bitcoin.conf) with `<IP>=manual-block-relay`
- don't participate in transaction and address relay
- don't get discouraged / punished for misbehavior (but will still get disconnected for sending TX/tx-INV as automatic block-relay-only peers do)
- are not subject to outbound eviction logic (unlike automatic block-relay-only-peers)

mzumsande committed:
p2p: add MANUAL_BLOCK_RELAY Connection type

This commit implements the p2p behaviour, but the new
connection type is not used yet.

mzumsande committed:
net, rpc: add block-relay-only option to addnode rpc

This applies to both add mode and onetry mode

Co-authored-by: dhruv <856960+dhruv@users.noreply.github.com>
Co-authored-by: brunoerg <brunoely.gc@gmail.com>

luke-jr commented on src/net_processing.cpp:
```diff
@@ -1273,7 +1273,7 @@ void PeerManagerImpl::FinalizeNode(const CNode& node)
     } // cs_main
     if (node.fSuccessfullyConnected && misbehavior == 0 &&
-        !node.IsBlockOnlyConn() && !node.IsInboundConn()) {
+        !node.IsBlockOnlyConn() && !node.IsManualBlockOnlyConn() && !node.IsInboundConn()) {
```
`IsBlockOnlyConn` and `IsManualConn` (below) should just return true for `MANUAL_BLOCK_RELAY` rather than additional checks every place

mzumsande replied:
I didn't do that because `IsBlockOnlyConn()` is used in several places (e.g. `PeerManagerImpl::EvictExtraOutboundPeers`, `CConnman::GetExtraBlockRelayCount()`, `GetCurrentBlockRelayOnlyConns()`) which would not apply to manual block-relay-only connections.

luke-jr commented on src/rpc/net.cpp:
```diff
@@ -278,6 +278,7 @@ static RPCHelpMan addnode()
                 {
                     {"node", RPCArg::Type::STR, RPCArg::Optional::NO, "The node (see getpeerinfo for nodes)"},
                     {"command", RPCArg::Type::STR, RPCArg::Optional::NO, "'add' to add a node to the list, 'remove' to remove a node from the list, 'onetry' to try a connection to the node once"},
+                    {"block_relay_only", RPCArg::Type::BOOL, RPCArg::Default{false}, "If set, treat as block-relay-only connection (deactivating transaction and addr relay on this connection)."},
```
This should be a `connection_type` String param, not a Boolean (see also #20551).

(Throw an exception if an unsupported value is passed, for this PR)

mzumsande replied:
Done with the latest push.

luke-jr commented on src/net.h:
```diff
@@ -877,7 +903,7 @@ class CConnman
     int GetExtraBlockRelayCount() const;
 
-    bool AddNode(const std::string& node);
+    bool AddNode(const std::string& node, bool block_relay_only);
```
Rather use `ConnectionType` here

mzumsande replied:
done

mzumsande commented:
I'd be really interested in opinions whether the use case (as outlined in OP and linked issue) is  important enough to justify the added complexity of adding another connection type (and also adding more complexity to `-addnode`).

mzumsande committed:
net, init: Add manual-block-relay option to bitcoind addnode

This allows to specify manual-block-relay peers per
command arg or bitcoin.conf

mzumsande commented:
Rebased and addressed feedback.

www222fff commented on src/net.h:
```diff
@@ -1087,7 +1114,7 @@ class CConnman
     std::deque<std::string> m_addr_fetches GUARDED_BY(m_addr_fetches_mutex);
     Mutex m_addr_fetches_mutex;
-    std::vector<std::string> m_added_nodes GUARDED_BY(m_added_nodes_mutex);
+    std::vector<AddedNodeEntry> m_added_nodes GUARDED_BY(m_added_nodes_mutex);
```
Why not try std::map<std::string, ConnectionType> m_added_nodes, then we can avoid to define AddedNodeEntry?

mzumsande replied:
Hmm, I'm a bit undecided about that: I think avoiding a struct is not really a good argument for switching form vector to map: We would need to switch to a pair or struct if we'd ever need to add another member, plus there is a case for introducing a struct even if we'd switch to a map, as was recently done in [PruneLocks](https://github.com/bitcoin/bitcoin/blob/037c5e511fe2185d244049cae25a98f99b878787/src/node/blockstorage.h#L139) example. 
Performance-wise, this doesn't seem important at all since the container cannot have more than 8 entries (MAX_ADDNODE_CONNECTIONS).

www222fff commented on src/net_processing.cpp:
```diff
@@ -1230,7 +1230,7 @@ void PeerManagerImpl::InitializeNode(CNode *pnode)
         assert(m_txrequest.Count(nodeid) == 0);
     }
-    PeerRef peer = std::make_shared<Peer>(nodeid, /*tx_relay=*/ !pnode->IsBlockOnlyConn());
+    PeerRef peer = std::make_shared<Peer>(nodeid, /*tx_relay=*/!pnode->IsBlockOnlyConn() && !pnode->IsManualBlockRelayConn());
```
Why not directly change IsBlockOnlyConn() directly as below, then can avoid to define new function IsManualBlockRelayConn()?

     bool IsBlockOnlyConn() const {
-        return m_conn_type == ConnectionType::BLOCK_RELAY;
+        return m_conn_type == ConnectionType::BLOCK_RELAY || m_conn_type == ConnectionType::MANUAL_BLOCK_RELAY;
     }

mzumsande replied:
See https://github.com/bitcoin/bitcoin/pull/24170#discussion_r795710002 - there are several placed that use `IsBlockOnlyConn()` with logic specific to automatic block-relay-only connections that have logic that does not apply to manual ones. So if I did that, we'd need additional logic in these places to distinguish between manual and automatic connections.

luke-jr replied:
That seems preferable?

mzumsande replied:
I have changed the logic now such that `IsBlockRelayConn` is used when both connection types are meant, where as `IsManualBlockRelayConn` and `IsAutomaticBlockRelayConn` are used in spots where only one is meant.

jnewbery commented on src/net.cpp:
```diff
@@ -2202,9 +2206,9 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
-    for (const std::string& strAddNode : lAddresses) {
-        CService service(LookupNumeric(strAddNode, Params().GetDefaultPort(strAddNode)));
-        AddedNodeInfo addedNode{strAddNode, CService(), false, false};
+    for (auto nodes_to_add : lAddresses) {
```
```suggestion
    for (auto node_to_add : lAddresses) {
```

(since you're iterating over the items individually, `node_to_add` is singular)

mzumsande replied:
done

jnewbery commented on src/net.cpp:
```diff
@@ -2202,9 +2206,9 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
-        AddedNodeInfo addedNode{strAddNode, CService(), false, false};
+    for (auto nodes_to_add : lAddresses) {
+        CService service(LookupNumeric(nodes_to_add.m_node_address, Params().GetDefaultPort(nodes_to_add.m_node_address)));
+        AddedNodeInfo addedNode{nodes_to_add.m_node_address, CService(), false, false, nodes_to_add.conn_type};
```
Since you're touching this line, can you comment the boolean arguments?

mzumsande replied:
done

jnewbery commented on src/net.h:
```diff
@@ -523,10 +541,16 @@ class CNode
         return m_conn_type == ConnectionType::INBOUND;
     }
 
+    bool IsManualBlockRelayConn() const
```
Since you're adding `IsManualBlockRelayConn()`, consider renaming `IsBlockOnlyConn()` to `IsAutomaticBlockRelayConn()`, otherwise that function name is misleading (it returns `false` for connections that are block relay only connections but were created manually).

mzumsande replied:
Done, see also https://github.com/bitcoin/bitcoin/pull/24170#discussion_r863023967.

jnewbery commented on src/net.h:
```diff
@@ -188,6 +180,31 @@ enum class ConnectionType {
+    ConnectionType conn_type;
+};
+
+struct AddedNodeEntry {
```
I know that this is maybe a bit of a stretch for this PR, but I really don't like the addnode/addednode terminology, and would much prefer this to be named something like `ManualConnectionEntry`, which much more precisely describes the nature of these connections.

mzumsande replied:
The downside of `ManualConnectionEntry` is that it is not a connection, just a name of a node we'll try to connect to (which may succeed or not). I didn't really like the name AddedNodeEntry either though, but couldn't think of anything better. Maybe `ManualNodeEntry`?

jnewbery replied:
Or maybe `ManualConnectionTarget` or `DesiredManualConnection`? I'm not sure they're great either.

jnewbery commented on src/net.h:
```diff
@@ -188,6 +180,31 @@ enum class ConnectionType {
+};
+
+struct AddedNodeEntry {
+    std::string m_node_address;
+    ConnectionType conn_type;
```
Be consistent with the `m_` prefix (either `m_node_address` and `m_conn_type` or `node_address` and `conn_type`). I prefer naming all members of structs with the `m_` prefix, but opinions vary.

mzumsande replied:
changed to `m_conn_type`

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -293,6 +294,20 @@ static RPCHelpMan addnode()
+    if (!request.params[2].isNull()) {
+        connection_type_name = request.params[2].get_str();
+    }
+    if (connection_type_name != "manual" && connection_type_name != "manual-block-relay") {
```
Is it possible to use `ConnectionTypeAsString()` here to avoid defining these strings in multiple places in the source?

mzumsande replied:
Good idea! Used `ConnectionTypeAsString()` here and in the last commit (also rearranged the rpc processing code a bit).

jnewbery reviewed (commented):
Concept ACK. This seems like a reasonable feature request.

I've left a few comments, mostly around naming.

I wonder if in future, we could change the net/net_processing responsibilities such that net_processing is unaware of whether a connection is manual or not. The only places that net_processing uses that information are in `MaybeDiscourageAndDisconnect()` (before this PR), and `EvictExtraOutboundPeers()` (in this PR). If disconnection and eviction logic is moved out of net_processing (cc @cfields), then net_processing won't need to be aware of the difference between manual and automatic connections.

luke-jr commented on src/init.cpp:
```diff
@@ -440,7 +440,7 @@ void SetupServerArgs(ArgsManager& argsman)
                  ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
 
-    argsman.AddArg("-addnode=<ip>", strprintf("Add a node to connect to and attempt to keep the connection open (see the addnode RPC help for more info). This option can be specified multiple times to add multiple nodes; connections are limited to %u at a time and are counted separately from the -maxconnections limit.", MAX_ADDNODE_CONNECTIONS), ArgsManager::ALLOW_ANY | ArgsManager::NETWORK_ONLY, ...
+    argsman.AddArg("-addnode=<ip>[=manual-block-relay]", strprintf("Add a node to connect to and attempt to keep the connection open (see the addnode RPC help for more info). This option can be specified multiple times to add multiple nodes; connections are limited to %u at a time and are counted separately from the -maxconnections limit. Append =manual-block-relay to disable transaction and addr ...
```
Prefer `=blocksonly` to be clearer and match the existing `-blocksonly` option.

jnewbery replied:
`-blocksonly` mode is different from block relay peers.

luke-jr replied:
How so? Maybe it needs better docs on the difference...

jnewbery replied:
Here you go: https://bitcoin.stackexchange.com/questions/112828/what-is-a-block-relay-only-connection-what-is-it-used-for

luke-jr replied:
That doesn't contrast.

mzumsande replied:
Some differences:
- -blocksonly is a global startup option (applying to all connections) -block-relay-only is connection-specific
- A node in -blocksonly mode participates in address relay, we never send out addresses on -block-relay-only connections (and ignore incoming ones).
- A node in -blocksonly signals not to receive any transactions (`fRelay`) and will disconnect if that is breached, but it may in special cases send out transactions itself (if submitted via RPC). We'll  never send a tx over a block-relay-only connection.
- A node in -blocksonly mode still has two of their outbound connections as -block-relay-only (for which then the block-relay-only rules apply).

www222fff commented on src/net.cpp:
```diff
@@ -2812,22 +2816,22 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
-    for (const std::string& it : m_added_nodes) {
-        if (strNode == it) return false;
+    for (const auto& it : m_added_nodes) {
+        if (strNode == it.m_node_address) return false;
```
Supposed option modify case need not be handled?
If add node firstly with setting manual and then setting manual block relay.

mzumsande replied:
No, a modify option is not supported, especially since it is not possible to change the connection connection type of an existing connection. A user could instead use the 'remove' option and then re-add with another connection type (which would not change existing connections though).

www222fff commented on src/net.cpp:
```diff
@@ -1258,6 +1260,7 @@ bool CConnman::AddConnection(const std::string& address, ConnectionType conn_typ
     switch (conn_type) {
     case ConnectionType::INBOUND:
     case ConnectionType::MANUAL:
+    case ConnectionType::MANUAL_BLOCK_RELAY:
```
Addconnection rpc should not support manual block relay, so seems here need not check?

mzumsande replied:
It's true that it doesn't support this, just as it doesn't support INBOUND or MANUAL. But leaving it out her would result in a compiler warning for the switch, see comment below:
`// no default case, so the compiler can warn about missing cases`

mzumsande committed:
scripted-diff: Rename IsBlockOnlyConn to IsAutomaticBlockRelayConn

This is in preparation for the addition of manual block-relay-only
connections.

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren IsBlockOnlyConn IsAutomaticBlockRelayConn
ren GetCurrentBlockRelayOnlyConns GetCurrentAutomaticBlockRelayConns

-END VERIFY SCRIPT-

mzumsande commented:
Thanks for the reviews @www222fff @luke-jr  @jnewbery I pushed an update, addressing the feedback.

jnewbery commented on src/net.h:
```diff
@@ -786,7 +815,9 @@ class CConnman
         {
             LOCK(m_added_nodes_mutex);
-            m_added_nodes = connOptions.m_added_nodes;
+            for (auto addedNode : connOptions.m_added_nodes) {
```
```suggestion
            for (auto added_node : connOptions.m_added_nodes) {
```

jnewbery commented on src/net.cpp:
```diff
@@ -2836,22 +2840,22 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
 }
 
-bool CConnman::AddNode(const std::string& strNode)
+bool CConnman::AddNode(const std::string& strNode, ConnectionType conn_type)
```
Should this assert or return false if `conn_type != ConnectionType::MANUAL && conn_type != ConnectionType::MANUAL_BLOCK_RELAY`? What would it mean to have a non-manual connection type in the `m_added_nodes` container?

If this is changed to an assert you'd need to also update the test in fuzz/connman.

mzumsande replied:
I added an assert and updated the fuzz test to comply with that.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -293,6 +294,22 @@ static RPCHelpMan addnode()
         throw std::runtime_error(
             self.ToString());
     }
+    std::string manual_name{ConnectionTypeAsString(ConnectionType::MANUAL)};
```
These can be const.

mzumsande replied:
done

jnewbery reviewed (commented):
Code review ACK 4ad26899169c45a6295efe8f1798e91ac0810e00

mzumsande commented:
[4ad2689 ](https://github.com/bitcoin/bitcoin/commit/4ad26899169c45a6295efe8f1798e91ac0810e00) to [6d75df7](https://github.com/bitcoin/bitcoin/commit/6d75df749d4aa9f2479fbe9d1551c8a5c1a3607f):
addressed feedback by @jnewbery

jnewbery commented:
Code review ACK 6d75df749d

jonatack commented:
It seems to me the linked issue would be solved in the general case by BIP342 (or by adding manual connections to other non-clearnet networks).

> In that situation, manual block-relay-only connections to peers on clearnet networks can help us staying connected to the best chain, but in contrast to normal manual connections, transactions and addresses aren't transmitted on these links - in particular not our own address or transaction from our wallet. This increases privacy and will also make it harder to perform fingerprinting attacks (connecting our identities over different networks).

Manual connections are to trusted peers, so if I understand correctly, the specific threat model here is that of a trusted peer fingerprinting us?

vasild commented:
> Manual connections are to trusted peers, so if I understand correctly, the specific threat model here is that of a trusted peer fingerprinting us?

My understanding is that the problem being addressed here is a passive MITM observing that a particular transaction originated from our node when we send it to our trusted peer, added with `-addnode` (over clearnet).

jonatack commented:
> > Manual connections are to trusted peers, so if I understand correctly, the specific threat model here is that of a trusted peer fingerprinting us?
> 
> My understanding is that the problem being addressed here is a passive MITM observing that a particular transaction originated from our node when we send it to our trusted peer, added with `-addnode` (over clearnet).

If so, my understanding was that BIP324 is designed to address this: https://drive.google.com/file/d/1o6PaA-vAWXhpCHHbBkUGZJ-6bmde6gEY/view

sipa commented:
Overall I think this is a useful addition, if it ends up being used. I'm a bit concerned it's too obscure for people to decide to configure links like this. Weak concept ACK.

@mzumsande / @dhruv It seems both this PR and #24545 add a way to associate flags with manually configured connections. Perhaps it's worth seeing if the interfaces can be made the same?

@vasild Maybe just a terminology note, but a MitM is by definition an active attacker that inserts themselves in the connection. I think perhaps you mean more something like an eavesdropper / large scale surveillance (which just observes without interfering)?

@jonatack Even BIP324 cannot fully prevent information about transaction origin to a global surveillant due to traffic patterns. We do intend to include affordances for dummy traffic / traffic shaping in BIP324, but actually making use of that to hide transaction propagation patterns is out of scope.

vasild commented:
@sipa, yes, by passive man in the middle, I mean eavesdropping or spying. Btw, [wikipedia MITM](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) reads "... attacker secretly relays and **possibly** alters the communications between two parties ..."

mzumsande commented:
> Manual connections are to trusted peers, so if I understand correctly, the specific threat model here is that of a trusted peer fingerprinting us?

There are different levels of trust and I think that it is quite typical that users would just choose a peer with a good uptime and no past misbehaviour as a "trusted" addnode peer without actually being able to vouch for its operator completely.

But even if the peer is indeed good, there is still the possibility of indirect attempts to infer this connection, where a third party would connect to us (on the privacy network where we accept connections) and also to our suspected addnode peer (via clearnet). In this situation, I think it would be possible to infer the existence of this connection with a reasonable degree of confidence using addr relay probing, and probably some tricks exist within tx relay as well. I think that there is a much larger attack surface if you participate in addr and tx relay, and part of the reason block-relay-only connections were introduced is to just disable these attack vectors completely on these connections instead of trying to solve all relevant issues with addr and tx relay.

> It seems both this PR and https://github.com/bitcoin/bitcoin/pull/24545 add a way to associate flags with manually configured connections. Perhaps it's worth seeing if the interfaces can be made the same?

Yes, this should be easily possible. I had a look at an older version of this (#23900), and I think that the changes in `net` are quite similar already. I will soon take a closer look and possibly change some nomenclature/details here to make them more compatible.

> I'm a bit concerned it's too obscure for people to decide to configure links like this.

Yes, that's definitely a valid point, I was unsure about this [myself](https://github.com/bitcoin/bitcoin/issues/23763#issuecomment-1098558732). It seems really hard to gauge what features users will actually use.

mzumsande commented:
6d75df7 to c348172:
Rebased and made the approach analogous to the one by @dhruv in [#24545](https://github.com/bitcoin/bitcoin/pull/24545/commits/36bb7117a45ee2e0c411ab1152b818e0acfa0c5c) by renaming `AddedNodeEntry` to `AddedNodeParams` and and making it a member of `AddedNodeInfo`.

mzumsande commented:
[c348172 ](https://github.com/bitcoin/bitcoin/commit/c34817279cca646be09d6210d2a7b3dd45b03fe0)to [8af5283](https://github.com/bitcoin/bitcoin/commit/8af528399289c89df5917aa1e7190265609b6eb2):

Rebased after the merge of https://github.com/bitcoin/bitcoin/pull/22778 - this meant adding a couple more calls to `IsBlockRelayConn()`, that used to be `m_tx_relay == nullptr` checks before  #22778 and therefore apply for both manual and automatic block-relay-only connections.

jnewbery commented:
reACK 8af5283992

jonatack commented:
Attempted a couple of week-long Twitter surveys related to this topic at https://twitter.com/jonatack/status/1524377548062351360 and https://twitter.com/jonatack/status/1524380125126660096.

LarryRuane commented:
> [6d75df7](https://github.com/bitcoin/bitcoin/commit/6d75df749d4aa9f2479fbe9d1551c8a5c1a3607f) to [c348172](https://github.com/bitcoin/bitcoin/commit/c34817279cca646be09d6210d2a7b3dd45b03fe0): Rebased and made the approach analogous to the one by @dhruv in [#24545](https://github.com/bitcoin/bitcoin/pull/24545/commits/36bb7117a45ee2e0c411ab1152b818e0acfa0c5c) by renaming `AddedNodeEntry` to `AddedNodeParams` and and making it a member of `AddedNodeInfo`.

Suggestion, if possible do the rebase and PR changes in separate force-pushes, to make it easier for reviewers to see just the PR changes. I think I've seen GitHub combine the two force-pushes if they're close in time, so it may be advisable to wait a bit between the force-pushes.

mzumsande commented:
[8af5283 ](https://github.com/bitcoin/bitcoin/commit/8af528399289c89df5917aa1e7190265609b6eb2)to [629a6f4](https://github.com/bitcoin/bitcoin/commit/629a6f404056e3228f49fd03e6c682d8f2820e37):
Rebased due to conflict with #25156.

> Suggestion, if possible do the rebase and PR changes in separate force-pushes, to make it easier for reviewers to see just the PR changes.

Ok, I'll try to think of that next time!

LarryRuane reviewed (commented):
ACK 629a6f404056e3228f49fd03e6c682d8f2820e37
comments are suggestions

mzumsande commented on src/net.cpp:
```diff
@@ -2227,9 +2231,9 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
+        CService service(LookupNumeric(node_to_add.m_node_address, Params().GetDefaultPort(node_to_add.m_node_address)));
+        AddedNodeInfo addedNode{node_to_add, CService(), /*fConnected=*/false, /*fInbound=*/false};
         if (service.IsValid()) {
             // strAddNode is an IP:port
```
I changed it to "node address", so I don't have to use a variable name here.

mzumsande commented on src/net.h:
```diff
@@ -785,7 +813,9 @@ class CConnman
         {
             LOCK(m_added_nodes_mutex);
-            m_added_nodes = connOptions.m_added_nodes;
+            for (auto added_node : connOptions.m_added_nodes) {
```
done, thanks.

mzumsande commented:
[629a6f4 ](https://github.com/bitcoin/bitcoin/commit/629a6f404056e3228f49fd03e6c682d8f2820e37)to [00ea370](https://github.com/bitcoin/bitcoin/commit/00ea3706617da8ff526e666661d98d339cda1372):
addressed suggestions by @LarryRuane - thank you!

LarryRuane commented on src/net.cpp:
```diff
@@ -3156,6 +3161,19 @@ uint64_t CConnman::CalculateKeyedNetGroup(const CAddress& address) const
     return GetDeterministicRandomizer(RANDOMIZER_ID_NETGROUP).Write(vchNetGroup.data(), vchNetGroup.size()).Finalize();
 }
 
+AddedNodeParams CConnman::GetAddedNodeParams(std::string added_node)
```
```suggestion
AddedNodeParams CConnman::GetAddedNodeParams(std::string added_node) const
```

mzumsande replied:
This became obsolete, since GetAddedNodeParams was removed.

LarryRuane commented on src/net.h:
```diff
@@ -1054,6 +1084,7 @@ class CConnman
 
     size_t SocketSendData(CNode& node) const EXCLUSIVE_LOCKS_REQUIRED(node.cs_vSend);
     void DumpAddresses();
+    AddedNodeParams GetAddedNodeParams(std::string addedNode);
```
```suggestion
    AddedNodeParams GetAddedNodeParams(std::string addedNode) const;
```

mzumsande replied:
This became obsolete, since GetAddedNodeParams was removed.

mzumsande committed:
doc: add release note for manual block-relay connections

mzumsande commented on src/net.cpp:
```diff
@@ -3156,6 +3161,19 @@ uint64_t CConnman::CalculateKeyedNetGroup(const CAddress& address) const
+        if (added_node.substr(index + 1) == ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)) {
+            added_node = added_node.substr(0, index);
+            conn_type = ConnectionType::MANUAL_BLOCK_RELAY;
+        }
```
I moved the check to init - in doing so, I realized that if I check for invalid connection types anyway, I might as well do the parsing of the string there as well - so I changed it to pass a vector of `AddedNodeParams` structs already in `CConnman::Options` and removed the CConnman::GetAddedNodeParams function. Let me know if that makes sense to you!

mzumsande commented:
Sorry, @LarryRuane  it took me some time to address your comments.
[00ea370 ](https://github.com/bitcoin/bitcoin/commit/00ea3706617da8ff526e666661d98d339cda1372)to [20b6e75](https://github.com/bitcoin/bitcoin/commit/20b6e75ddc88a8d11aaae32700e606b5eada24f5):
- Addressed feedback to validate the connection type when passing via bitcoind arg. Note that the node name is not being validated, like before.
- Added a commit with a relase note.

I also extended the test coverage of the RPC -addnode to cover manual-block-relay connections (`rpc_net.py`).

jnewbery commented on doc/release-notes.md:
```diff
@@ -104,6 +104,11 @@ Low-level changes
+- The RPC `addnode` and the `-addnode` bitcoind startup option now
+optionally support adding manual peers that are block-relay-only.
+On these connections, the node doesn't participate in transaction or
+address relay. (#24170)
```
Perhaps add "see bitcoind help documentation and RPC help documentation for parameter semantics" or similar

mzumsande replied:
Done, in a slightly shortened form.

jnewbery reviewed (commented):
utACK 20b6e75ddc88a8d11aaae32700e606b5eada24f5

Happy to re-review once this has been rebased.

mzumsande commented:
[20b6e75 ](https://github.com/bitcoin/bitcoin/commit/20b6e75ddc88a8d11aaae32700e606b5eada24f5) to [191aea3](https://github.com/bitcoin/bitcoin/commit/191aea32ab77893d9e97a2792851a16e8d463bd7): rebased
(will address comments with next push)

mzumsande commented on src/net_processing.cpp:
```diff
@@ -413,7 +413,7 @@ struct CNodeState {
       * Any peer protected (m_protect = true) is not chosen for eviction. A peer is
       * marked as protected if all of these are true:
-      *   - its connection type is IsBlockOnlyConn() == false
+      *   - its connection type is IsAutomaticBlockRelayConn() == false
```
Good point, I changed this in the commit where the manual-block-relay connections are introduced.

mzumsande commented:
[d37c267 ](https://github.com/bitcoin/bitcoin/commit/d37c267503bb94d2e93a8b3c01a4afd444337af8)to [70595a5](https://github.com/bitcoin/bitcoin/commit/70595a585483539b19e3fe77f07e32d06063266b):
rebased (conflict with #25591)

jnewbery commented:
utACK 70595a585483539b19e3fe77f07e32d06063266b

Verified git range-diff

mzumsande commented:
[70595a5 ](https://github.com/bitcoin/bitcoin/commit/70595a585483539b19e3fe77f07e32d06063266b)to [690b040 ](https://github.com/bitcoin/bitcoin/commit/690b040c6e011e4ec1195dde9932cdf3f367ff5d):
rebased due to conflict with #25514

mzumsande commented:
[690b040 ](https://github.com/bitcoin/bitcoin/commit/690b040c6e011e4ec1195dde9932cdf3f367ff5d)to [894486a](https://github.com/bitcoin/bitcoin/commit/894486a7420303f77010b9c49cd4633deb2fd00a):
rebased (conflict with #25699)

mzumsande commented:
[894486a ](https://github.com/bitcoin/bitcoin/commit/894486a7420303f77010b9c49cd4633deb2fd00a)to [9bc3178](https://github.com/bitcoin/bitcoin/commit/9bc317883358fb27851de50c256f36505f2fc51a): 

Rebased (moved release note into its own file)

brunoerg commented:
Concept ACK

Should we consider the manual block-relay-only connections as anchors? E.g I added 2 manual block-relay-only connections in addition to the 2 outbound block-relay-only ones and I'd like to consider the manual ones as my anchors. so, when my node starts it would connect to that 2 peers. Could it make sense?

mzumsande commented:
Rebased

> Should we consider the manual block-relay-only connections as anchors? E.g I added 2 manual block-relay-only connections in addition to the 2 outbound block-relay-only ones and I'd like to consider the manual ones as my anchors. so, when my node starts it would connect to that 2 peers. Could it make sense?

This PR allows to add the manual block-relay-only connections as addnode peers to `bitcoin.conf` - I think if we do this, it's essentially the same behavior as if it was an anchor, we'll connect to them with each new restart.

willcl-ark reviewed (approved):
ACK 3eb8c42368a6e059b1842cbcec478815d947a341

If you touch again (to kick the CI.) you might consider also amending [this comment](https://github.com/bitcoin/bitcoin/blob/3eb8c42368a6e059b1842cbcec478815d947a341/src/net_processing.cpp#L2725-L2726) to clarify that that manual-block-relay peers are **not** excluded from the protection, only automatic-block-relay peers are?

aureleoules commented on src/init.cpp:
```diff
@@ -1713,7 +1713,23 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        const size_t index = added_node.rfind('=');
+        if (index != std::string::npos) {
+            const std::string conn_name = added_node.substr(index + 1);
+            if (conn_name == ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)) {
+                added_node = added_node.substr(0, index);
+                conn_type = ConnectionType::MANUAL_BLOCK_RELAY;
+            } else {
```
Maybe also add the case for `ConnectionType::MANUAL`?

`./src/bitcoind -addnode='test.local=manual'`
> 2023-01-16T16:04:19Z Error: Invalid -addnode connection type: manual
> Error: Invalid -addnode connection type: manual

mzumsande replied:
Not sure about this: Currently, the help for "addnode" states that only `=manual-block-relay` is possible (manual is the default). Adding this to the help might only complicate things for the user, because there would be no difference between `=manual` and no argument at all?

brunoerg replied:
I prefer not having `=manual`, it would add more stuff to the code and I believe it might complicate for the user, it's the default behavior, not sure why someone would want to specify it.

aureleoules reviewed (approved):
Light ACK 3eb8c42368a6e059b1842cbcec478815d947a341

The code looks good and I verified that the tests correctly test the new behavior.
Be wary I am not familiar with net processing.

rajarshimaitra reviewed (commented):
Concept + tACK 3eb8c42368a6e059b1842cbcec478815d947a341

I think this is a reasonable addition, though share the concern about how much it would actually be used in public. Mostly seems to be not hurting much, modulo the code complexity added due to a new type.

Below are a few nonblocking comments and questions.

achow101 commented:
Any net people want to take a look here?

jonatack commented:
As mentioned in the issue #23763, perhaps I'm missing something but this seems somewhat redundant with

- using a VPN, and/or
- running bitcoind over multiple encrypted networks (Tor, I2P, CJDNS), and/or
- BIP324 v2 encrypted p2p in the future

achow101 commented:
> As mentioned in the issue #23763, perhaps I'm missing something but this seems somewhat redundant with
> 
> * using a VPN, and/or
> * running bitcoind over multiple encrypted networks (Tor, I2P, CJDNS), and/or

I don't see how any of those are relevant. Privacy networks and VPNs are generally bandwidth constrained and have a much smaller pool of nodes to connect to, therefore making it easier for an attacker to eclipse you. Being able to add manual blocksonly connections allows nodes to add connections to clearnet nodes for public data that they are unlikely to originate - blocks. Any transactions would still be relayed through connections made on the privacy networks so their origins are masked.

> * BIP324 v2 encrypted p2p in the future

Since we're talking about active attackers, BIP 324 doesn't help.

jonatack commented:
> Privacy networks and VPNs are generally bandwidth constrained

I've been running bitcoind over VPNs and all of our encrypted networks since years from without bandwidth issues, and currently since 3 months from rural Central America on slow/flakey connections.  Even when GitHub takes a long time to load a page, bitcoind thankfully runs well in making peer connections under adverse conditions from what I can see between the debug log and a live -netinfo 4 dashboard.

> and have a much smaller pool of nodes to connect to, therefore making it easier for an attacker to eclipse you

As one example, my little dev node on my laptop out here currently knows ~14k Tor peers and ~1k I2P peers after the `IsTerrible` filter.  15k seems like a pretty decent amount of peers to connect to.

<img width="368" alt="Screenshot 2023-01-31 at 10 50 37" src="https://user-images.githubusercontent.com/2415484/215854831-40623357-1b34-432f-98db-7bbc0f6e31e2.png">

LarryRuane commented on src/net.cpp:
```diff
@@ -2523,22 +2525,23 @@ std::vector<CAddress> CConnman::GetAddresses(CNode& requestor, size_t max_addres
     }
 
-    m_added_nodes.push_back(strNode);
+    m_added_nodes.push_back({strNode, conn_type});
```
```suggestion
    m_added_nodes.push_back(AddedNodeParams{
        .m_node_address{strNode},
        .m_conn_type{conn_type}});

```

mzumsande replied:
done

LarryRuane commented on src/net.h:
```diff
@@ -972,7 +988,7 @@ class CConnman
     /**
      * Return vector of current BLOCK_RELAY peers.
```
Doing this rename would touch 20 lines of code in 8 files, but could be done with a scripted diff. I agree it would be more consistent, so personally, I'm in favor, but I'd also understand not wanting to increase the diff that much.

mzumsande replied:
I agree that it would be more consistent and added a commit to do this - I'll check how many additional conflicts this introduces...

LarryRuane commented on src/rpc/net.cpp:
```diff
@@ -299,6 +300,25 @@ static RPCHelpMan addnode()
+    if (strCommand == "remove" && !request.params[2].isNull()) {
+        throw JSONRPCError(RPC_INVALID_PARAMS, "Cannot specify a connection type with 'remove'");
+    }
+    std::string connection_type_name = manual_name;
```
```suggestion
    std::string connection_type_name{manual_name};
```

LarryRuane commented on src/rpc/net.cpp:
```diff
@@ -299,6 +300,25 @@ static RPCHelpMan addnode()
+    ConnectionType connection_type;
+    if (connection_type_name == manual_block_relay_name){
+        connection_type = ConnectionType::MANUAL_BLOCK_RELAY;
+    }
+    else if (connection_type_name == manual_name) {
```
clang-format says (also below)
```suggestion
    } else if (connection_type_name == manual_name) {
```

mzumsande replied:
done

LarryRuane reviewed (commented):
re-ACK 9e953f1134fe4b752868f519f12d8dee5c1cef52

Code suggestions are advisory.

Suggestion for PR description, maybe include an RPC (or config) example so reviewers can try it out manually quickly without having to look at the code or help to figure out the syntax, something like
```
bitcoin-cli addnode <IP>:8333 add manual-block-relay
```

The results of manual testing look good. After enabling logging category "net" and adding a manual-block-relay peer, this appears in `debug.log`:
```
New outbound peer connected: version: 70016, blocks=775913, peer=24 (manual-block-relay)
```
After that, the only logging for peer=24 is sending and receiving ping and pong, and also:
```
[net] received: headers (82 bytes) peer=24
[net] received: addrv2 (44 bytes) peer=24
[net] ignoring addrv2 message from manual-block-relay peer=24
```

mzumsande commented:
I was busy with other things in the last weeks, but I will get back to address feedback later this week!

mzumsande committed:
scripted-diff: Rename IsOutboundOrBlockRelayConn to IsAutomaticOutboundOrBlockRelayConn

to make it clearer that manual connections are not included

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren IsOutboundOrBlockRelayConn IsAutomaticOutboundOrBlockRelayConn

-END VERIFY SCRIPT-

mzumsande committed:
scripted-diff: rename BLOCK_RELAY to AUTOMATIC_BLOCK_RELAY

to better distinguish from MANUAL_BLOCK_RELAY

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren BLOCK_RELAY AUTOMATIC_BLOCK_RELAY

-END VERIFY SCRIPT-

mzumsande commented on src/net.h:
```diff
@@ -424,6 +428,7 @@ class CNode
             case ConnectionType::MANUAL:
             case ConnectionType::ADDR_FETCH:
             case ConnectionType::FEELER:
+            case ConnectionType::MANUAL_BLOCK_RELAY:
                 return false;
```
It's used in `net_processing` in places suchas  `ConsiderEviction()` where we wouldn't want to consider manual connections. 
I think the name "IsOutboundOrBlockRelayConn" isn't ideal (even before this PR, because manual connections are  also outbound connections), maybe the name "IsAutomaticOutboundOrBlockRelayConn" would be better. I added a commit to do that.

mzumsande commented on test/functional/rpc_net.py:
```diff
@@ -218,13 +218,25 @@ def test_getaddednodeinfo(self):
+        assert_raises_rpc_error(-5, "Unsupported connection type", self.nodes[0].addnode, node=dummy_node, command='add', connection_type='foo')
+        # add a manual-block-relay node (the validity of the 'node' parameter is not validated)
+        self.nodes[0].addnode(node=dummy_node, command='add', connection_type='manual-block-relay')
+        added_nodes = self.nodes[0].getaddednodeinfo(dummy_node)
+        assert_equal(len(added_nodes), 1)
+        assert_equal(added_nodes[0]['addednode'], dummy_node)
+        assert_equal(added_nodes[0]['connection_type'], 'manual-block-relay')
+        self.nodes[0].addnode(node=dummy_node, command='remove')
```
Yes, I think it makes sense as a follow-up! Though, if I'm not missing something we don't even support regular MANUAL connections there, so adding support for these would make sense as well?

mzumsande commented on src/net.cpp:
```diff
@@ -1899,20 +1901,20 @@ std::vector<AddedNodeInfo> CConnman::GetAddedNodeInfo() const
-        AddedNodeInfo addedNode{strAddNode, CService(), false, false};
+    for (const auto& node_to_add : lAddresses) {
+        CService service(LookupNumeric(node_to_add.m_node_address, Params().GetDefaultPort(node_to_add.m_node_address)));
+        AddedNodeInfo addedNode{node_to_add, CService(), /*fConnected=*/false, /*fInbound=*/false};
```
I think I prefer the more concise version here, especially when most of the args are just defaults that are set to their actual values in the following lines.

mzumsande commented:
Sorry for the delay - I addressed the outstanding comments and pushed an update.

Major changes are two scripted-diff commits renaming the function.
`IsOutboundOrBlockRelayConn` to `IsAutomaticOutboundOrBlockRelayConn`
and the connection type
`BLOCK_RELAY` to `AUTOMATIC_BLOCK_RELAY`


> If you touch again (to kick the CI.) you might consider also amending [this comment](https://github.com/bitcoin/bitcoin/blob/3eb8c42368a6e059b1842cbcec478815d947a341/src/net_processing.cpp#L2725-L2726) to clarify that that manual-block-relay peers are **not** excluded from the protection, only automatic-block-relay peers are?

Since the protection only applies to Full Outbound connections (and nothing else), manual block-relay peers are also excluded from it (just like all other manual peers). It just makes most sense naming (automatic) block-relay-only peers here.

> Not sure if I was clear, just to clarify. If I use -addnode=ip=manual-block-relay in CLI, not putting it in bitcoin.conf, when I shutdown my node, this peer is not gonna be dump into anchors.dat, right?

yes, that is correct, anchors are only chosen from automatic block-relay-only connections. I think it makes sense not to mix automatic and manual connections that way: 
If the node operator wants to be permanently connected to a given manual connection, they can add this to their permanent config.

mzumsande commented:
Sorry for the delay
[b5b6300 ](https://github.com/bitcoin/bitcoin/commit/b5b630099b24f5a87d316c9e5af258543fb329eb)to [bd56522](https://github.com/bitcoin/bitcoin/commit/bd56522466df3a899bea4ff86bfa2afc8af5049c)
I addressed the outstanding comments and pushed an update. Also had to rebase.

Major changes are two additional scripted-diff commits renaming the function.
`IsOutboundOrBlockRelayConn` to `IsAutomaticOutboundOrBlockRelayConn`
and the connection type
`BLOCK_RELAY` to `AUTOMATIC_BLOCK_RELAY`


> If you touch again (to kick the CI.) you might consider also amending [this comment](https://github.com/bitcoin/bitcoin/blob/3eb8c42368a6e059b1842cbcec478815d947a341/src/net_processing.cpp#L2725-L2726) to clarify that that manual-block-relay peers are **not** excluded from the protection, only automatic-block-relay peers are?

Since the protection only applies to Full Outbound connections (and nothing else), manual block-relay peers are also excluded from it (just like all other manual peers). It just makes most sense naming (automatic) block-relay-only peers here.

> Not sure if I was clear, just to clarify. If I use -addnode=ip=manual-block-relay in CLI, not putting it in bitcoin.conf, when I shutdown my node, this peer is not gonna be dump into anchors.dat, right?

yes, that is correct, anchors are only chosen from automatic block-relay-only connections. I think it makes sense not to mix automatic and manual connections that way: 
If the node operator wants to be permanently connected to a given manual connection, they can add this to their permanent config.

mzumsande commented:
> As mentioned in the issue #23763, perhaps I'm missing something but this seems somewhat redundant with
> 
> * using a VPN, and/or
> * running bitcoind over multiple encrypted networks (Tor, I2P, CJDNS), and/or
> * BIP324 v2 encrypted p2p in the future

I'm not only thinking about a third party monitoring your connections, but also about about the information you'd give the clearnet outbound peers you connect to.
I think there are good reasons to have a connection to clearnet (getting blocks quickly, eclipse attack resistance), and a way to do that which ensures you don't send your wallet transactions over these connections makes sense to me  
(especially since rebroadcast still has privacy issues).

For me, the critical question is still if this opt-in feature will find enough users to justify the added complexity in `net` - I honestly don't know that, and I wouldn't have a problem closing this for now if people think there isn't enough demand for this use case currently.

stratospher commented:
> For me, the critical question is still if this opt-in feature will find enough users to justify the added complexity in net - I honestly don't know that, and I wouldn't have a problem closing this for now if people think there isn't enough demand for this use case currently.

if it helps, i found some related issues: (but i don’t know this either)

- there’s an [issue in raspiblitz](https://github.com/rootzoll/raspiblitz/issues/3329) where they talk about wanting to connect their tor node to clearnet to speed up IBD and and not lose privacy. they ended up using i2p peers.
- there’s also [https://github.com/bitcoin/bitcoin/issues/19042](https://github.com/bitcoin/bitcoin/issues/19042) and [https://github.com/bitcoin/bitcoin/issues/14692](https://github.com/bitcoin/bitcoin/issues/14692) where they talk about wanting to connect to clearnet peers without relaying own transactions.

in the very unlikely scenario that everyone likes this connection and all bitcoin nodes on tor connect with clearnet only using this option - that wouldn't be good for the network right? (assuming an average tor user wouldn’t want their addresses/transactions relayed to their clearnet peers)

mzumsande commented:
> in the very unlikely scenario that everyone likes this connection and all bitcoin nodes on tor connect with clearnet only using this option - that wouldn't be good for the network right? (assuming an average tor user wouldn’t want their addresses/transactions relayed to their clearnet peers)

We'd need enough inbound capacity on clearnet to accomodate the additional manual connections to it, and also some users relaying transactions on both networks, otherwise txns submitted through tor wouldn't make it into clearnet and vice versa. I think that there are many nodes with a lesser need for privacy (e.g. because they don't submit transactions regularly) that don't have the need for something like and would be fine having full connections to both networks.

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -303,6 +304,23 @@ static RPCHelpMan addnode()
         throw std::runtime_error(
             self.ToString());
     }
+    const std::string manual_name{ConnectionTypeAsString(ConnectionType::MANUAL)};
```
nit: you could create `manual_name` and `manual_block_relay_name` after checking the connection type with "remove":
```diff
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 6155af323..9d87d89e3 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -304,11 +304,14 @@ static RPCHelpMan addnode()
         throw std::runtime_error(
             self.ToString());
     }
-    const std::string manual_name{ConnectionTypeAsString(ConnectionType::MANUAL)};
-    const std::string manual_block_relay_name{ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)};
+
     if (strCommand == "remove" && !request.params[2].isNull()) {
         throw JSONRPCError(RPC_INVALID_PARAMS, "Cannot specify a connection type with 'remove'");
     }
+
+    const std::string manual_name{ConnectionTypeAsString(ConnectionType::MANUAL)};
+    const std::string manual_block_relay_name{ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)};
+
     std::string connection_type_name{manual_name};
     if (!request.params[2].isNull()) {
         connection_type_name = request.params[2].get_str();
```

mzumsande replied:
made obsolete by the next comment.

brunoerg commented on src/rpc/net.cpp:
```diff
@@ -303,6 +304,23 @@ static RPCHelpMan addnode()
+    if (strCommand == "remove" && !request.params[2].isNull()) {
+        throw JSONRPCError(RPC_INVALID_PARAMS, "Cannot specify a connection type with 'remove'");
+    }
+    std::string connection_type_name{manual_name};
```
perhaps there is a way to simplify it:
```diff
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 6155af323..b9e082a63 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -304,22 +304,19 @@ static RPCHelpMan addnode()
         throw std::runtime_error(
             self.ToString());
     }
-    const std::string manual_name{ConnectionTypeAsString(ConnectionType::MANUAL)};
-    const std::string manual_block_relay_name{ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)};
+
     if (strCommand == "remove" && !request.params[2].isNull()) {
         throw JSONRPCError(RPC_INVALID_PARAMS, "Cannot specify a connection type with 'remove'");
     }
-    std::string connection_type_name{manual_name};
+
+    const std::string manual_block_relay_name{ConnectionTypeAsString(ConnectionType::MANUAL_BLOCK_RELAY)};
+    ConnectionType connection_type{ConnectionType::MANUAL};
     if (!request.params[2].isNull()) {
-        connection_type_name = request.params[2].get_str();
-    }
-    ConnectionType connection_type;
-    if (connection_type_name == manual_block_relay_name) {
-        connection_type = ConnectionType::MANUAL_BLOCK_RELAY;
-    } else if (connection_type_name == manual_name) {
-        connection_type = ConnectionType::MANUAL;
-    }  else {
-        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Unsupported connection type: " + connection_type_name);
+        const std::string connection_type_name{request.params[2].get_str()};
+        if (connection_type_name == manual_block_relay_name)
+            connection_type = ConnectionType::MANUAL_BLOCK_RELAY;
+        else
+            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Unsupported connection type: " + connection_type_name);
     }
 
     NodeContext& node = EnsureAnyNodeContext(request.context);
```

mzumsande replied:
Yes, thanks, that's a bit simpler and should be equivalent, so I took your suggestion!

mzumsande commented:
b36db71 to c13c986: rebased (will address the oustanding review comments soon!)

mzumsande commented on src/net.h:
```diff
@@ -123,6 +115,18 @@ struct CSerializedNetMsg {
+
+struct AddedNodeInfo {
+    AddedNodeParams m_params;
+    CService resolvedAddress;
```
`m_node_address` is a string, which is part of AddedNodeParams, which has just the unvalidated user input.

`resolvedAddress` is a `CService` built from `m_node_address`, which is generated in `GetAddedNodeInfo` and used in `ThreadOpenConnection` / rpc code.

mzumsande commented:
[c13c986 ](https://github.com/bitcoin/bitcoin/commit/c13c9864c8f50a74dfca8feba42e2d28831decaf)to [41ce312](https://github.com/bitcoin/bitcoin/commit/41ce31289fdb521d3f3af1b67f750ce4a9196391):
Incorporated feedback by @brunoerg

mzumsande commented:
Seems like other proposals such as #27213 and #27509 that rely on improving automatic behavior have more support compared to this one (which needs manual setup by the node operator). Closing for now.
